server/src/main/java/org/opensearch/gateway/ReplicaShardBatchAllocator.java

```diff
@@ -51,13 +53,27 @@ public abstract class ReplicaShardBatchAllocator extends ReplicaShardAllocator {
      */
     public void processExistingRecoveries(RoutingAllocation allocation, List<List<ShardRouting>> shardBatches) {
         List<Runnable> shardCancellationActions = new ArrayList<>();
+
+        Set<ShardId> initReplicasFromRouting = allocation.routingNodes()
+            .shardsWithState(ShardRoutingState.INITIALIZING)
+            .stream()
+            .filter(r -> !r.primary())
+            .map(ShardRouting::shardId)
```

Comment:
can we keep ShardRouting itself here? Can there be case where for shard ShardRouting object is different in RoutingAllocation vs batch cache?

Reply:
This might be a problem in case of multiple replica copies for a shard. RoutingAllocation would have one ShardRouting per copy whereas the batch cache would contain only one of the ShardRouting (picked randomly).

ShardRouting might not exactly match and hence de-dupe by ShardId is required.

Reply:
But then fetching and decision should happen on latest ShardRouting from RoutingAllocation instead of Batch cache

Reply:
fixed this to consider all the ShardRoutings for a replica